### PR TITLE
Add APK support to Ahab

### PR DESCRIPTION
**Pull request by DarthHater**
@ken-duck has added Alpine linux to OSS Index (soon anyways), so we can add apk support to `ahab`

This pull request makes the following changes:
* New APK files that do all the fun processing
* Regex stolen from `nexus-repository-apk` that parses version and name from an apk filename
* New test, and new `testapk.txt` to test this

This can be tested like so:

`cat testapk.txt | ./ahab chase --os alpine` or by running `apk info -vv | sort |./ahab chase --os alpine` if you run it in Alpine Linux

cc @bhamail / @DarthHater / @zendern

### CODE REVIEW

File: parse/apk.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+	splitPackage := strings.Split(pkg, " ")
+	parsedProject.Name = doApkShowParseName(splitPackage[0])
+	parsedProject.Version = doApkShowParseVersion(splitPackage[0])
+	return
+}
+
+func doApkShowParseName(pkg string) (name string) {
+	re, err := regexp.Compile("([a-zA-Z][a-zA-Z0-9]+-)+")
+	if err != nil {
+		fmt.Println(err)
```
Comment (zendern) [inline parse/apk.go]: If the regex really fails to compile it basically borked right?? Would it make more sense to just panic and let the program die?? It would point out in development that something is broken and hopefully make the feedback loop faster in theory ¯\_(ツ)_/¯

File: parse/apk.go [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+	if results != nil {
+		name = strings.TrimSuffix(results[0], "-")
+	}
+	return
+}
+
+func doApkShowParseVersion(pkg string) (version string) {
+	re, err := regexp.Compile("([0-9]+)(\\.[0-9]+)(\\.[0-9]+)?(-[r][0-9]+)")
+	if err != nil {
+		fmt.Println(err)
```
Comment (zendern) [inline parse/apk.go]: Same goes for this one

File: parse/apk.go [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/sonatype-nexus-community/nancy/types"
+)
+
+func ParseApkShow(packages []string) (projectList types.ProjectList) {
+	for _, pkg := range packages {
+		if strings.Contains(pkg, "WARNING") {
```
Comment (zendern) [inline parse/apk.go]: Preference thing but maybe just negate this and remove the empty if block.

File: parse/apk_test.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+ssl_client-1.30.1-r2 - EXternal ssl_client for busybox wget
+zlib-1.2.11-r1 - A compression/decompression Library`
+
+var apkShowArray = strings.Split(apkShowList, "\n")
+
+func TestParseApkShowList(t *testing.T) {
+	result := ParseApkShow(apkShowArray)
+
+	// adduser 3.116ubuntu1
+	assert.Equal(t, types.Projects{"alpine-baselayout", "3.1.2-r0"}, result.Projects[0])
```
Comment (zendern) [inline parse/apk_test.go]: Any reason not to just combine these 2 tests?? If I'm looking at them correctly they are testing the same method and just passing in different data. 

Why not just add a length check in this second test and nix the first one??

File: testapk.txt [hunk 5]
```diff
@@ -0,0 +1,16 @@
+WARNING: Ignoring APKINDEX.00740ba1.tar.gz: No such file or directory
```
Comment (zendern) [inline testapk.txt]: In PR #7 I started down the path of having tests that actually spun up docker containers and validated it did what it should. It's cool that this file was added and the comment in the PR on how to use it but realistically after this PR is merged this file is just a thing that is not used other than in manual testing (maybe).

File: parse/apk.go (hunk 1 above)
Comment (DarthHater) [reply parse/apk.go]: That is a really great point!

File: parse/apk.go (hunk 2 above)
Comment (DarthHater) [reply parse/apk.go]: In: d121e08

File: parse/apk.go (hunk 1 above)
Comment (DarthHater) [reply parse/apk.go]: In: d121e08

File: parse/apk_test.go (hunk 4 above)
Comment (DarthHater) [reply parse/apk_test.go]: In: 698e51f

File: parse/apk.go (hunk 3 above)
Comment (DarthHater) [reply parse/apk.go]: In: c313d9c

File: testapk.txt (hunk 5 above)
Comment (DarthHater) [reply testapk.txt]: I like where you are going, it would be nice to replace these with real tests!

File: parse/apk_test.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+
+// generate CLI package list via:
+// # apk info -vv | sort
+var apkShowList = `WARNING: Ignoring APKINDEX.00740ba1.tar.gz: No such file or directory
+WARNING: Ignoring APKINDEX.d8b2a6f4.tar.gz: No such file or directory
+alpine-baselayout-3.1.2-r0 - Alpine base dir structure and init scripts
+alpine-keys-2.1-r2 - Public keys for Alpine Linux packages
+apk-tools-2.10.4-r2 - Alpine Package Keeper - package manager for alpine
+busybox-1.30.1-r2 - Size optimized toolbox of many common UNIX utilities
+ca-certificates-cacert-20190108-r0 - Mozilla bundled certificates
```
Comment (bhamail) [inline parse/apk_test.go]: There may be a boog lurking here. The `ca-certificates-cacert-20190108-r0` case produced an empty version string, where I think it should produce `20180409-r0`

e.g.

    // ca-certificates 20180409
	assert.Equal(t, types.Projects{"ca-certificates", "20180409-r0"}, result.Projects[4])

Not sure it helps, but here's the regex I was using when I started to implement APK. Mine did not use `apk info -vv`, but only `apk info -v`, so less verbose. I was also removing the "-rN" suffix because I was not sure how ossindex would handle versions with the "-rN" suffix. Different inputs, so not sure if it helps much, but just in case...

```
func doApkParse(pkg string) (parsedProject types.Projects) {
	pkg = strings.TrimSpace(pkg)

	re, err := regexp.Compile(`^((.*)-([^a-zA-Z].*)-.*)`)
	if err != nil {
		fmt.Println(err)
	}
	newSlice := re.FindStringSubmatch(pkg)
	if newSlice != nil {
		parsedProject.Name = newSlice[2]
		parsedProject.Version = newSlice[3]
	} else {
		// yikes, nothing we recognize.
		fmt.Printf("filaure parseing namve,version for package: %s\n", pkg)
	}
	return
}
```

File: parse/apk_test.go [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+		t.Errorf("Didn't work, expected %d projects but got %d", 14, len(result.Projects))
+	}
+
+	// adduser 3.116ubuntu1
+	assert.Equal(t, types.Projects{"alpine-baselayout", "3.1.2-r0"}, result.Projects[0])
+
+	// apt 1.6.12
+	assert.Equal(t, types.Projects{"alpine-keys", "2.1-r2"}, result.Projects[1])
+
+	// ca-certificates 20180409
```
Comment (bhamail) [inline parse/apk_test.go]: minor: mismatched comment from prior edge case test

File: parse/apk_test.go (hunk 6 above)
Comment (DarthHater) [reply parse/apk_test.go]: In: 6e673b8

File: parse/apk_test.go (hunk 7 above)
Comment (DarthHater) [reply parse/apk_test.go]: In: 6e673b8

File: parse/apk_test.go [hunk 8]
```diff
@@ -50,12 +50,12 @@ func TestParseApkShowList(t *testing.T) {
...
-	// apt 1.6.12
-	assert.Equal(t, types.Projects{"alpine-keys", "2.1-r2"}, result.Projects[1])
+	// alpine-keys-2.1-r2
+	assert.Equal(t, types.Projects{"alpine-keys", "2.1"}, result.Projects[1])
 
-	// ca-certificates 20180409
-	assert.Equal(t, types.Projects{"apk-tools", "2.10.4-r2"}, result.Projects[2])
+	// apk-tools-2.10.4-r2
+	assert.Equal(t, types.Projects{"apk-tools", "2.10.4"}, result.Projects[2])
 }
```
Comment (bhamail) [inline parse/apk_test.go]: Would you mind adding an assertion for the edge case that exposed the version parsing issue? e.g.

```
ca-certificates-cacert-20190108-r0
```

Comment (DarthHater) [reply parse/apk_test.go]: Para ti: e951b38

**Comment by DarthHater** [comment]
@zendern @bhamail I think I'm going to merge this, any objections?

**Comment by zendern** [comment]
@darthhater is the apk stuff official in the index now?? I’d say do it if it is.

**Comment by bhamail** [comment]
+1 no objections from me
